Add dbt Cloud provider

josh-fell opened the pull request:
This PR adds a new provider to interface with [dbt Cloud](https://docs.getdbt.com/docs/dbt-cloud/cloud-overview) via the [dbt Cloud API](https://docs.getdbt.com/dbt-cloud/api-v2).  The provider includes:

- `DbtCloudHook` which implements an abstraction for almost all of the available endpoints in the dbt Cloud API and inherits from the existing `HttpHook`.
- Two operators, `DbtCloudRunJobOperator` and `DbtCloudGetJobRunArtifactOperator`, which triggers a dbt Cloud job and downloads a run artifact, respectively.
- A sensor, `DbtCloudJobRunSensor`, to poll status of a specific dbt Cloud job run.
- An operator link to allow users to navigate directly to the triggered dbt Cloud job run for monitoring.
- A `test_connection()` method in the `DbtCloudHook` for users to test connections prior to executing DAGs.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

josh-fell committed:
Add dbt Cloud provider

josh-fell committed:
Add docstrings throughout

josh-fell committed:
Add more operator args and default trigger_reason

josh-fell committed:
Cleanup markers for rst exampleinclude

josh-fell committed:
Add pagination to DbtCloudHook

josh-fell committed:
Updates, both functional and typing, from unit testing

josh-fell committed:
Add unit tests for DbtCloudHook

josh-fell committed:
Add unit tests for DbtCloudJobRunSensor

josh-fell committed:
Comment fixes to sensor unit test

josh-fell committed:
Add minor test cases for DbtCloudHook

josh-fell committed:
Small refactor for DbtCloudHook tests

josh-fell committed:
DRY refactor of DbtCloudHook tests - part 2

josh-fell committed:
Add unit tests for operators

josh-fell committed:
Remove :type: directives, update example DAG, + Mypy fixes

josh-fell committed:
Update assertion call in test_list_job_runs

josh-fell committed:
Add connection and operator docs

josh-fell committed:
Fix static check errors

josh-fell committed:
Trivial comment, docstring, and logging updates

mik-laj commented on airflow/providers/dbt/cloud/operators/dbt.py:
```diff
@@ -0,0 +1,203 @@
+    ) -> None:
+        super().__init__(**kwargs)
+        self.dbt_cloud_conn_id = dbt_cloud_conn_id
+        self.hook = DbtCloudHook(self.dbt_cloud_conn_id)
```
We should avoid initializing hook in ctor.

josh-fell replied:
Right on. I'll pull this initialization out.

mik-laj commented on docs/apache-airflow-providers-dbt-cloud/connections.rst:
```diff
@@ -0,0 +1,66 @@
+    .. note::
+
+      If an Account ID is not provided in an Airflow connection, ``account_id`` *must* be explicitly passed to
+      an operator or hook method.
```
Can you add exaample URI as bash commnd?
For exmaple - Google URL
```
export AIRFLOW_CONN_GOOGLE_CLOUD_SQL_DEFAULT='gcpcloudsql://user:XXXXXXXXX@1.1.1.1:3306/mydb?database_type=mysql&project_id=example-project&location=europe-west1&instance=testinstance&use_proxy=True&sql_proxy_use_tcp=False'
```
https://airflow.apache.org/docs/apache-airflow-providers-google/stable/connections/gcp_sql.html

josh-fell replied:
Absolutely. Will add.

josh-fell committed:
Move DbtCloudHook() out of constructor + update conn doc

mik-laj commented on docs/apache-airflow-providers-dbt-cloud/connections.rst:
```diff
@@ -0,0 +1,87 @@
+
+    .. code-block:: bash
+
+        export AIRFLOW_CONN_DBT_CLOUD_DEFAULT='dbt-cloud://api_token@'
```
```suggestion
        export AIRFLOW_CONN_DBT_CLOUD_DEFAULT='dbt-cloud://:api_token@'
```

mik-laj commented on docs/apache-airflow-providers-dbt-cloud/connections.rst:
```diff
@@ -0,0 +1,87 @@
+
+        export AIRFLOW_CONN_DBT_CLOUD_DEFAULT='dbt-cloud://api_token@'
+
+More information on creating connections via environment variables can be found
```
Please use meaningful link text. See: https://developers.google.com/style/cross-references

josh-fell replied:
Oh thanks for the tip on this reference. Really nice documentation on best practices.

josh-fell committed:
Fix conn doc URI example and improve link text

subkanthi commented on airflow/providers/dbt/cloud/hooks/dbt.py:
```diff
@@ -0,0 +1,488 @@
+        :return: The request response.
+        """
+        return self._run_and_get_response(
+            endpoint=f"{account_id}/runs/{run_id}/",
```
Should we log an error if account_id is not passed, looks like it might construct the url with None.

josh-fell replied:
This should be caught by the `@fallback_to_default_account` decorator which wraps every hook method that calls a dbt Cloud API endpoint directly. The decorator checks to see if the `account_id` is not passed or if it is, if the value is `None`. If either of those conditions are met, the `account_id` set in the Airflow connection is used. In the event there is no `account_id` configured in the connection, this exception is raised: `raise AirflowException("Could not determine the dbt Cloud account.")`. Does that suffice?

subkanthi commented on airflow/providers/dbt/cloud/hooks/dbt.py:
```diff
@@ -0,0 +1,488 @@
+                if num_current_results < num_total_results:
+                    response = self.run(endpoint=endpoint, data=_paginate_payload)
+                    resp_json = response.json()
+                    if resp_json["data"]:
```
would it help to have a log  statement in the "else" if data is not present in the response.

josh-fell replied:
Certainly couldn't hurt. Perhaps this is better as a debug statement though since it's more related to the aspects of the API response rather than the task execution.

josh-fell replied:
I removed this check. The `data` object has no bearing on pagination. Thanks for the catch!

josh-fell committed:
Remove "data" key check during pagination

josh-fell committed:
Fix static check errors after rebase

sungchun12 commented on airflow/providers/dbt/cloud/hooks/dbt.py:
```diff
@@ -0,0 +1,487 @@
+    def __call__(self, request: PreparedRequest) -> PreparedRequest:
+        request.headers["Content-Type"] = "application/json"
+        request.headers["Authorization"] = f"Token {self.token}"
+
```
Add in a user agent to the header, and dynamically pass in the version of the airflow dbt Cloud provider to track usage on the dbt Cloud side. 

```suggestion
        request.headers["User-Agent"] = f"airflow-dbt-cloud-provider-v{X.Y.Z}"
```

sungchun12 replied:
Note: the committed code is different from this suggestion!

josh-fell committed:
Update operator link to not create ad hoc TaskInstance

See #21285

josh-fell committed:
Add user-agent header with provider version info

sungchun12 reviewed (approved):
Thanks for adding the change Josh. I verified it's tracking on the dbt Cloud side in our logs!

josh-fell committed:
Updated cached property name

Having a `cached_property` named "conn" while there is a `get_conn()` method of which both don't return the same thing could be confusing.

potiuk reviewed (approved):
I think it's a great DBT provider start. Just in time for Feb release.
